File: src/app/services/MaintenanceTicket.service.ts

import { Injectable } from '@angular/core';
import { FormGroup,  FormBuilder,  Validators } from '@angular/forms';
import { HttpClient } from '@angular/common/http';
import { Observable, Subject } from 'rxjs';
import {MaintenanceTicket} from '../models/MaintenanceTicket';
import {IoTDeviceService} from '../services/IoTDevice.service';
import {TenantService} from '../services/Tenant.service';
import { HelperBaseService } from './helperbase.service';

@Injectable({
	providedIn: 'root'
})

export class MaintenanceTicketService extends HelperBaseService {

	//********************************************************************
	// general holder 
	//********************************************************************
	maintenanceTicket : MaintenanceTicket;

	//********************************************************************
	// Catch all for the return value of a service call
	//********************************************************************
	result: any;

	//********************************************************************
	// sole constructor, injected with the HttpClient
	//********************************************************************
	constructor(private http: HttpClient) {
		super();
	}

		//********************************************************************
	// add a MaintenanceTicket
	// returns the results untouched as a JSON representation
	// delegates via URI
	//********************************************************************
	addMaintenanceTicket(ticketNumber, openedAt, closedAt, Device, Tenant, Priority, Status) : Observable<any> {
		const uri_ = this.apiUrl + '/MaintenanceTicket/create';
		const obj = {
			      		ticketNumber: ticketNumber,
      		openedAt: openedAt,
      		closedAt: closedAt,
      		Device: Device != null && Device.length > 0 ? Device : null,
      		Tenant: Tenant != null && Tenant.length > 0 ? Tenant : null,
      		Priority: Priority,
			Status: Status
		};

		return this.http.post(uri_, obj);
	}

	//********************************************************************
	// update a MaintenanceTicket
	// returns an Observable
	// delegates via URI
	//********************************************************************
		updateMaintenanceTicket(ticketNumber, openedAt, closedAt, Device, Tenant, Priority, Status, id)  :  Observable<any>  {
			const uri_ = this.apiUrl + '/MaintenanceTicket/update/' + id;
		const obj = {
				      		ticketNumber: ticketNumber,
      		openedAt: openedAt,
      		closedAt: closedAt,
      		Device: Device != null && Device.length > 0 ? Device : null,
      		Tenant: Tenant != null && Tenant.length > 0 ? Tenant : null,
      		Priority: Priority,
			Status: Status
		};
		return this.http.post(uri_, obj);
	}

	//********************************************************************
	// delete a MaintenanceTicket
	// returns an Observable
	// delegates via URI
	//********************************************************************
	deleteMaintenanceTicket(id)  : Observable<any> {
		const uri_ = this.apiUrl + '/MaintenanceTicket/delete/' + id;

		return this.http.get(uri_);
	}
	
	//********************************************************************
	// loads a MaintenanceTicket
	// returns the results untouched as an Observable MaintenanceTicket
	// MaintenanceTicket model
	// delegates via URI
	//********************************************************************
	getMaintenanceTicket(id) : Observable<MaintenanceTicket> {
		const uri_ = this.apiUrl + '/MaintenanceTicket/load/' + id;

		return this.http.get<MaintenanceTicket>(uri_);
	}
	
	//********************************************************************
	// gets all MaintenanceTicket
	// returns the results untouched as JSON representation of an
	// Observable array of MaintenanceTicket models
	// delegates via URI
	//********************************************************************
	getMaintenanceTickets() : Observable<MaintenanceTicket[]> {
		const uri_ = this.apiUrl + '/MaintenanceTicket/';

		return this
			.http.get<MaintenanceTicket[]>(uri_);
	}
	
		
	//********************************************************************
	// assigns a Device on a MaintenanceTicket
	// returns an Observable
	// delegates via URI
	//********************************************************************
	assignDevice( maintenanceTicketId, _deviceId ): Observable<any> {

		// get the MaintenanceTicket from storage
		this.loadHelper( maintenanceTicketId );

	// get the IoTDevice from storage
	var tmp 	= new IoTDeviceService(this.http).getIoTDevice(_deviceId);

	// assign the Device
	this.maintenanceTicket.device = tmp;

	// save the MaintenanceTicket
	return this.saveHelper();
}

	//********************************************************************
	// unassigns a Device on a MaintenanceTicket
	// returns an Observable
	// delegates via URI
	//********************************************************************
	unassignDevice( maintenanceTicketId ): Observable<any> {

		// get the MaintenanceTicket from storage
		this.loadHelper( maintenanceTicketId );

	// assign Device to null
	this.maintenanceTicket.device = null;

	// save the MaintenanceTicket
	return this.saveHelper();
}

	
	//********************************************************************
	// assigns a Tenant on a MaintenanceTicket
	// returns an Observable
	// delegates via URI
	//********************************************************************
	assignTenant( maintenanceTicketId, _tenantId ): Observable<any> {

		// get the MaintenanceTicket from storage
		this.loadHelper( maintenanceTicketId );

	// get the Tenant from storage
	var tmp 	= new TenantService(this.http).getTenant(_tenantId);

	// assign the Tenant
	this.maintenanceTicket.tenant = tmp;

	// save the MaintenanceTicket
	return this.saveHelper();
}

	//********************************************************************
	// unassigns a Tenant on a MaintenanceTicket
	// returns an Observable
	// delegates via URI
	//********************************************************************
	unassignTenant( maintenanceTicketId ): Observable<any> {

		// get the MaintenanceTicket from storage
		this.loadHelper( maintenanceTicketId );

	// assign Tenant to null
	this.maintenanceTicket.tenant = null;

	// save the MaintenanceTicket
	return this.saveHelper();
}

	
	
	//********************************************************************
	// saveHelper - internal helper to save a MaintenanceTicket
	//********************************************************************
	saveHelper() : Observable<any> {

		const uri_ = this.apiUrl + '/MaintenanceTicket/update/' + this.maintenanceTicket;

	return  this.http.post(uri_, this.maintenanceTicket );
}

	//********************************************************************
	// loadHelper - internal helper to load a MaintenanceTicket
	//********************************************************************	
	loadHelper( id ) {
		this.getMaintenanceTicket(id)
			.subscribe((res : MaintenanceTicket) => {
				this.maintenanceTicket = res;
			});
	}
}

File: iotonaspdotnet/Service/MaintenanceTicketService.cs
using iotonaspdotnet.Domain;
using iotonaspdotnet.Persistence;
using iotonaspdotnet.Contracts;

namespace iotonaspdotnet.Service;

public interface IMaintenanceTicketService {

    Task Create(MaintenanceTicket model , CancellationToken cancellationToken);
    Task<bool> Update(MaintenanceTicket model, CancellationToken cancellationToken);
    Task<MaintenanceTicket?> Get(IdentifierRequest identifier, CancellationToken cancellationToken);
    Task<IReadOnlyList<MaintenanceTicket>> GetAll(CancellationToken cancellationToken);
    Task<bool> Delete(IdentifierRequest identifier, CancellationToken cancellationToken);

    // ------------------------------
    // Single Associations
    // -------------------------------
    Task<bool> AssignDevice(AssociationRequest request, CancellationToken cancellationToken);
    Task<bool> UnassignDevice(AssociationRequest request, CancellationToken cancellationToken);
    Task<bool> AssignTenant(AssociationRequest request, CancellationToken cancellationToken);
    Task<bool> UnassignTenant(AssociationRequest request, CancellationToken cancellationToken);


}

public class MaintenanceTicketService : IMaintenanceTicketService
{
    private readonly IMaintenanceTicketRepository _repository;

    public MaintenanceTicketService(
        IMaintenanceTicketRepository repository )
    {
        _repository = repository;
    }


    public async Task Create(MaintenanceTicket model, CancellationToken cancellationToken)
    {

 
         await _repository.AddAsync(model, cancellationToken);
    }

    public async Task<bool> Update(MaintenanceTicket model, CancellationToken cancellationToken)
    {
        var existing = await _repository.GetByIdAsync(model.Id, cancellationToken);
        if (existing is null)
        {
            return false;
        }
        existing.TicketNumber = model.TicketNumber;
        existing.OpenedAt = model.OpenedAt;
        existing.ClosedAt = model.ClosedAt;
        existing.Priority = model.Priority;
        existing.Status = model.Status;

        await _repository.UpdateAsync(existing, cancellationToken);
        return true;
    }

    public Task<MaintenanceTicket?> Get(IdentifierRequest identifier, CancellationToken cancellationToken)
    => _repository.GetByIdAsync(identifier.Id, cancellationToken);

    public Task<IReadOnlyList<MaintenanceTicket>> GetAll(CancellationToken cancellationToken)
    => _repository.GetAllAsync(cancellationToken);

    public async Task<bool> Delete(IdentifierRequest identifier, CancellationToken cancellationToken)
    {
        var existing = await _repository.GetByIdAsync(identifier.Id, cancellationToken);
        if (existing is null)
        {
            return false;
        }

        await _repository.DeleteAsync(existing, cancellationToken);
        return true;
    }

    public async Task<bool> AssignDevice(AssociationRequest request, CancellationToken cancellationToken) {
        return true;
    }
    public async Task<bool> UnassignDevice(AssociationRequest request, CancellationToken cancellationToken) {
        return true;
    }

    public async Task<bool> AssignTenant(AssociationRequest request, CancellationToken cancellationToken) {
        return true;
    }
    public async Task<bool> UnassignTenant(AssociationRequest request, CancellationToken cancellationToken) {
        return true;
    }




}
